ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala

```diff
@@ -162,31 +163,85 @@ class JdbcIndexerFactory[Status <: InitStatus] private (loggerFactory: NamedLogg
 class JdbcIndexer private[index] (
     initialInternalOffset: Long,
     beginAfterExternalOffset: Option[LedgerString],
-    ledgerDao: LedgerDao)(implicit mat: Materializer)
+    ledgerDao: LedgerDao,
+    metricsManager: MetricsManager)(implicit mat: Materializer)
     extends Indexer
     with AutoCloseable {
 
   @volatile
   private var headRef = initialInternalOffset
 
+  @volatile
+  private var lastReceivedRecordTime = Instant.now()
+
+  @volatile
+  private var lastReceivedOffset: LedgerString = _
```

Comment:
An alternative to a volatile variable would be to stamp the metrics with every time a new update is processed. 

I guess it doesn't make much difference because in that case we would be hitting similar synchronization mechanism (potentially a lot more expensive) inside of the metrics subsystem?

Reply:
So the offset is reported as a `Gauge`, which can represent any value really (ultimately `toString` will be called on it). Any other sort of metric is number based, which means we would have to convert the offset (array[long]) to a single number. Therefore I think using the gauge is more appropriate.
Disclaimer: the value of the gauge is not actively pushed. it is up to the configured metrics reporter to poll the gauge at some interval.

Reply:
What you put into the disclaimer was what I was mostly interested in. 

I was wondering about the performance implications of this approach to gathering metrics, i.e. volatile var + poll. But then concluded myself that there is no free lunch and what you have done is probably as good as it gets. "Active pushing" or a "volatile var" both exert some form of protection of the state shared between the threads. Just wanted to confirm that this is also what you think.

Reply:
oh, the polling happens in any case, because usually the reporter takes a snapshot of all metrics after a configured interval and sends the data to graphite, datadog, influxdb, ...